app/src/test/java/com/codesoom/assignment/infra/UserRepositoryTest.java

```diff
@@ -0,0 +1,125 @@
+package com.codesoom.assignment.infra;
+
+import static com.codesoom.assignment.constants.UserConstants.ID;
+import static com.codesoom.assignment.constants.UserConstants.NAME;
+import static com.codesoom.assignment.constants.UserConstants.EMAIL;
+import static com.codesoom.assignment.constants.UserConstants.PASSWORD;
+import static com.codesoom.assignment.constants.UserConstants.USER;
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.Optional;
+
+import com.codesoom.assignment.domain.User;
+import com.codesoom.assignment.domain.UserRepository;
+
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
+
+@DataJpaTest
+@DisplayName("UserRepository 클래스")
+public class UserRepositoryTest {
+    @Autowired
+    UserRepository userRepository;
+
+    private User user;
+
+    private User subjectSave() {
+        return userRepository.save(USER);
+    }
+
+    private void subjectDelete() {
+        userRepository.delete(user);
+    }
+
+    private Optional<User> subjectFindById() {
+        return userRepository.findById(user.getId());
+    }
+
+    @Nested
+    @DisplayName("save 메서드는")
+    public class Describe_save {
+        @AfterEach
+        private void afterEach() {
+            subjectDelete();
+        }
+
+        @Test
+        @DisplayName("User를 저장한다.")
+        public void it_saves_users() {
+            user = subjectSave();
+
+            assertThat(subjectFindById())
+                .isPresent();
+        }
+    }
```

Comment:
테스트를 잘 작성하셨어요. 하지만 좀 더 디테일을 향상시킬 수 있습니다. 전역변수 `user`의 사용을 최대한 줄이고 다음과 같은 모양으로 테스트할 수 있도록 수정해 보세요. 실험실을 확실히 격리시켜서 깨끗한 테스트를 만드는 것이라 할 수 있습니다. 지금 코드를 읽어보면 `subjectFindById()`에서 좋지 못한 냄새가 납니다. 어떤 아이디를 가져오는지 확인하려면 메소드 구현을 찾아 읽어야 하는 문제도 있지요.

```suggestion
        @Test
        @DisplayName("User를 저장한다.")
        public void it_saves_users() {
            final User user = subjectSave();

            assertThat(subjectFindById(user.getId()))
                .isPresent();
        }
    }
```

Reply:
테스트 대상을 완전히 격리시키는것이 subject 메서드의 용도인 것으로 생각하였습니다.
따라서 어떤 값도 리턴하지 않고, 어떤 인자도 받지 않도록 하였습니다.
조언을 듣고 보니, 테스트 대상의 메서드 시그네처를 변경하더라도, subject 메서드 시그네처가 변경될 일은 매우 적겠군요.
착각했던것 같습니다. 조언 감사합니다.

Reply:
이해하시면서 이해한 과정을 설명해주시는 것이 대웅님의 좋은 습관인 것 같습니다.

테스트 대상을 `subject` 내에 캡슐화시켜서, 테스트 대상 메소드의 이름이나 시그니처가 바뀌어도 테스트 코드의 일부만 수정해도 되도록 하는 것이 `subject`의 목표 중 하나라 할 수 있습니다. 테스트의 중요한 목표 중 하나는 테스트 대상의 리턴값을 검증하는 것이니 리턴값이 자연스럽게 흘러나오도록 해주는 것이 바람직하죠.